File: python/Generate Parentheses.py
class Solution(object):
    answer = []
    def dfs(self, left, right, paren):
        if left > right or left < 0 or right < 0:
            return
        
        if left == 0 and right == 0:
            self.answer.append(paren)
        
        self.dfs(left - 1, right, paren + '(')
        self.dfs(left, right - 1, paren + ')')
        
    def generateParenthesis(self, n):
        self.answer = []
        self.dfs(n, n, )
        return self.answer

File: cpp/Generate Parentheses.cpp
class Solution {
public:
    vector<string> answer;
    vector<string> generateParenthesis(int n) {
        helper(, n, n);
        return answer;
    }
    
    void helper(string str, int left, int right) {
        if (right < left) {
            return;
        }
        if (left == 0 && right == 0) {
            answer.push_back(str);
        }
        
        if (left > 0) {
            helper(str + '(', left - 1, right);
        }
        if (right > 0) {
            helper(str + ')', left, right - 1);
        }
    }
};
